hips/hip-00xx-sequencing.md

```diff
@@ -0,0 +1,192 @@
+---
+hip: 9999
+title: "Better Support for Resource Creation Sequencing"
+authors: [ "Joe Beck <joebeck5705@gmail.com>", "Evans Mungai <mbuevans@gmail.com>" ]
+created: "2024-11-15"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+This HIP is to propose a new featureset in Helm 4 to provide Application developers, who create heam charts for their applications, a well supported way of defining what order chart resources and chart dependencies should be deployed to Kubernetes.
+
+## Motivation
+
+Today, to accomplish resource sequencing you have two options. The first is leveraging helm hooks, the second is building required sequencing into the application (via startup code or init containers). The existing hooks and weights can be tedious to build and maintain for Application developers, and built-in app sequencing can unecessarily increase complexity of a Helm application that needs to be maintained by Application Developers. Helm as a package manager should be capable of enabling developers to properly sequence how their applications are deployed, and by providing such a mechanism to developers, this will significantly improve the Application developer's experience.
+
+Additionally, Helm currently doesn't provide a way to sequence when chart dependencies are deployed, and this featureset would ideally address this.
+
+## Rationale
+
+### Proposal 1: Named Dependencies
+This design was chosen due to simplicity and clarity of how it works for both the Helm developer and Application Operator
+
+### Proposal 2: Weighted Resouces (without hooks)
+This design was proposed as an alternative that functions very similarly to hooks today, to provide the same sequencing functionality outside of the scope of hooks.
+
+## Specification
+
+### Proposal 1: Named Dependencies
+At a high level, allow Chart Developers to assign named dependency annotations to both their Helm templated resources and Helm chart dependencies that Helm then uses to generate a deployment sequence at installation time. Three new annotations would be added to enable this described next.
+
+- `helm.sh/layer`: Declare a layer that a given resource belongs to. Any numebr of resources can belong to a layer.
+- `helm.sh/depends-on/layer`: Declare a layer that must exist and in a ready state before this resource can be deployed.
+- `helm.sh/depends-on/chart`: Declare a chart dependency that must exist and in a ready state before this resource can be deployed. For the chart to be declared ready, all of its resources, with their sequencing order taken into consideration, would need to be deployed and declared ready.
```

Comment:
To be clear, `helm.sh/depends-on/chart` refers to some chart that is not listed as a dependency of the current chart? Is that right?
- assuming this because the wait flag should already check that dependent charts are fully installed before installing the current chart
- is the idea that this is limited to other charts that have already been deployed to the same namespace, or is there an idea to check charts in other namespaces (assuming the user has correct RBAC permissions)?

My other question here is, do we want to allow depending on other named k8s resources that are not defined by a chart? Eg, `helm.sh/depends-on/resource`? Or is the idea that users could annotate other resources not defined by a chart with `helm.sh/layer`, declaring that they are considered part of the same layer of a chart?

Reply:
`helm.sh/depends-on/chart` is referring to any direct chart dependencies. I was unsure if Helm waits to install the current chart until all dependent charts are installed and assumed not. If Helm is already waiting for dependent charts to be installed, then this section can probably be removed?

Reply:
> assuming this because the wait flag should already check that dependent charts are fully installed before installing the current chart

`helm.sh/depends-on/chart` meant for both dependencies and subcharts, will wait regardless of `--wait/--wait-for-jobs` CLI flags being present. Unless we want to have a way to opt-out from sequencing, which we can then assume if `--wait/--wait-for-jobs` are absent, not sequencing. Today, `--wait` will wait for certain resources only. In the proposal we have `helm.sh/resource-ready` which a chart developer can use to define what they mean by "readiness". 

I think we need to add a section explaining how `--wait/--wait-for-jobs` will work with sequencing annotations. I added it to **Open Issues**

> My other question here is, do we want to allow depending on other named k8s resources that are not defined by a chart? Eg, helm.sh/depends-on/resource? Or is the idea that users could annotate other resources not defined by a chart with helm.sh/layer, declaring that they are considered part of the same layer of a chart?

I think using `helm.sh/depends-on/resource` to specify a resource makes sense. Helm would eventually interpret layers as a group of resources to wait for

Reply:
A suggestion came my way regarding reasoning around sequencing chart dependencies/subcharts. Chart authors would want to place such dependencies on a chart-level definition rather than inside chart resources. It makes sense because only one reference of the chart dependency is needed. Any other references are repetition since helm will wait for the chart before deploying the parent chart. We came up with something like below

In the parent `Chart.yaml`
```yaml
name: MyChart
annotations:
  # List of charts that need to be installed before MyChart
  helm.sh/depends-on/charts: "subchart, minio, rabbitmq"
dependencies:
- name: rabbitmq
  version: "1.2.3"
- name: minio
  version: "5.0.0"
```